[Ripple] Make rippleRadius a property

codeman7 opened the pull request:
This change modifies the rippleRadius _ivar_ to make it a _property_. This exposes it in the private header of the `MDCRippleLayer.h` so that later if we need to modify the rippleRadius to support min/max values we can. This change also allows us to test this property.

codeman7 committed:
Make rippleRadius a property

material-automation commented:
bazel detected changes to the following targets:

```
//components/AppBar:snapshot_tests
//components/AppBar:snapshot_tests_test_bundle
//components/AppBar:unit_tests
//components/AppBar:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/AppBar:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/AppBar:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/AppBar:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/AppBar:unit_tests_IPHONE_X_IN_11_0
//components/AppBar:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/Banner:snapshot_tests
//components/Banner:snapshot_tests_test_bundle
//components/BottomAppBar:snapshot_tests
//components/BottomAppBar:snapshot_tests_test_bundle
//components/BottomAppBar:unit_tests
//components/BottomAppBar:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/BottomAppBar:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/BottomAppBar:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/BottomAppBar:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/BottomAppBar:unit_tests_IPHONE_X_IN_11_0
//components/BottomAppBar:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/ButtonBar:snapshot_tests
//components/ButtonBar:snapshot_tests_test_bundle
//components/ButtonBar:unit_tests
//components/ButtonBar:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/ButtonBar:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/ButtonBar:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/ButtonBar:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/ButtonBar:unit_tests_IPHONE_X_IN_11_0
//components/ButtonBar:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/Buttons:snapshot_tests
//components/Buttons:snapshot_tests_test_bundle
//components/Buttons:unit_tests
//components/Buttons:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Buttons:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Buttons:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Buttons:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Buttons:unit_tests_IPHONE_X_IN_11_0
//components/Buttons:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/Cards:snapshot_tests
//components/Cards:snapshot_tests_test_bundle
//components/Cards:unit_tests
//components/Cards:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Cards:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Cards:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Cards:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Cards:unit_tests_IPHONE_X_IN_11_0
//components/Cards:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/Chips:snapshot_tests
//components/Chips:snapshot_tests_test_bundle
//components/Chips:unit_tests
//components/Chips:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Chips:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Chips:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Chips:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Chips:unit_tests_IPHONE_X_IN_11_0
//components/Chips:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/Dialogs:snapshot_tests
//components/Dialogs:snapshot_tests_test_bundle
//components/Dialogs:unit_tests
//components/Dialogs:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Dialogs:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Dialogs:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Dialogs:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Dialogs:unit_tests_IPHONE_X_IN_11_0
//components/Dialogs:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/NavigationBar:snapshot_tests
//components/NavigationBar:snapshot_tests_test_bundle
//components/NavigationBar:unit_tests
//components/NavigationBar:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/NavigationBar:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/NavigationBar:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/NavigationBar:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/NavigationBar:unit_tests_IPHONE_X_IN_11_0
//components/NavigationBar:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/Ripple:snapshot_tests
//components/Ripple:snapshot_tests_test_bundle
//components/Ripple:unit_tests
//components/Ripple:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Ripple:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Ripple:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Ripple:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Ripple:unit_tests_IPHONE_X_IN_11_0
//components/Ripple:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/ShadowElevations:snapshot_tests
//components/ShadowElevations:snapshot_tests_test_bundle
//components/Snackbar:snapshot_tests
//components/Snackbar:snapshot_tests_test_bundle
//components/Snackbar:unit_tests
//components/Snackbar:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Snackbar:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Snackbar:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Snackbar:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Snackbar:unit_tests_IPHONE_X_IN_11_0
//components/Snackbar:unit_tests_IPHONE_X_IN_11_0_test_bundle
```

<!-- identifier: affected-targets -->

material-automation commented on components/Ripple/src/private/MDCRippleLayer.m:
```diff
@@ -40,14 +38,14 @@ - (void)setNeedsLayout {
-      CGRectMake(CGRectGetMidX(self.bounds) - _rippleRadius,
-                 CGRectGetMidY(self.bounds) - _rippleRadius, _rippleRadius * 2, _rippleRadius * 2);
+      CGRectMake(CGRectGetMidX(self.bounds) - self.rippleRadius,
+                 CGRectGetMidY(self.bounds) - self.rippleRadius, self.rippleRadius * 2, self.rippleRadius * 2);
```
clang-format suggested the following change:


```
-  CGRect ovalRect =
-      CGRectMake(CGRectGetMidX(self.bounds) - self.rippleRadius,
-                 CGRectGetMidY(self.bounds) - self.rippleRadius, self.rippleRadius * 2, self.rippleRadius * 2);
+  CGRect ovalRect = CGRectMake(CGRectGetMidX(self.bounds) - self.rippleRadius,
+                               CGRectGetMidY(self.bounds) - self.rippleRadius,
+                               self.rippleRadius * 2, self.rippleRadius * 2);
```

<!-- identifier: clang-format -->
<!-- File: components/Ripple/src/private/MDCRippleLayer.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

codeman7 committed:
clang formatting

yarneo reviewed (dismissed):
As a user, how would I set the rippleRadius if it is not exposed in MDCRippleView? My expectation is that user's won't try to manipulate the ripple layer. furthermore, the ripple layers are usually transient and sometimes there are multiple layers appearing at once. By changing one ripple layer, it wouldn't necessarily mean you have altered the radius of the ripple layer you intended.

Also, I would try not to rely on `setNeedsLayout` not being called after you set a custom ripple radius. There should be some mechanism that if a custom ripple radius is set, it would always look at that value and not reset it to the default every time setNeedsLayout is called.

codeman7 committed:
Address feedback

yarneo committed:
Update MDCRippleLayer.m

codeman7 committed:
Update test

material-automation commented on components/Ripple/src/private/MDCRippleLayer.m:
```diff
@@ -28,26 +28,24 @@
+  CGFloat radius = self.rippleRadius > 0 ? self.rippleRadius : DefaultRippleRadius(self.bounds);
+  CGRect ovalRect = CGRectMake(CGRectGetMidX(self.bounds) - radius,
+                               CGRectGetMidY(self.bounds) - radius,
+                               radius * 2, radius * 2);
```
clang-format suggested the following change:


```
-                               CGRectGetMidY(self.bounds) - radius,
-                               radius * 2, radius * 2);
+                               CGRectGetMidY(self.bounds) - radius, radius * 2, radius * 2);
```

<!-- identifier: clang-format -->
<!-- File: components/Ripple/src/private/MDCRippleLayer.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

codeman7 committed:
clang

jverkoey commented on components/Ripple/src/private/MDCRippleLayer.m:
```diff
@@ -28,26 +28,23 @@
-                 CGRectGetMidY(self.bounds) - _rippleRadius, _rippleRadius * 2, _rippleRadius * 2);
+  CGFloat radius = self.rippleRadius > 0 ? self.rippleRadius : DefaultRippleRadius(self.bounds);
+  CGRect ovalRect = CGRectMake(CGRectGetMidX(self.bounds) - radius,
+                               CGRectGetMidY(self.bounds) - radius, radius * 2, radius * 2);
```
It looks like `self.rippleRadius` can affect `self.path`, but changing the value of `self.rippleRadius` doesn't appear to result in any logical path that would cause `self.path` to be updated as part of this change. Is this expected?

codeman7 replied:
Good catch! I missed the setter for rippleRadius. I added behavior to cover this instance and test in [f0a34b4](https://github.com/material-components/material-components-ios/pull/7350/commits/f0a34b4deb1f56cda2cc6c133a8729f26f73694b).

yarneo replied:
Becuase `[self setPathFromRadii];` is called every time before the ripple animation starts, is there a reason why you would need it in the setter too?

The ripple layers themselves are usually transient (unless in selected or tap is held), if setting the ripple layer alters `self.path` live, then it would affect the existing ripple and cause unwanted behavior. Ideally, setting the rippleRadius will affect future animations and not the existing layer. Therefore I would probably remove this setter, unless you think I'm missing something.

codeman7 replied:
I missed the call to `startRippleAtPoint...` calling `setPathFromRadii`. I don't think a setter is needed. @jverkoey Does that make sense?

jverkoey replied:
Fair points @yarneo. @codeman7 let's document the points @yarneo brought up in the API's docs so that another engineer doesn't accidentally make this change in the future.

Specifically:

```
maxRippleRadius will only affect new ripples. If a ripple is already being animated when this property is changed,
that ripple's radius will not be affected.
```

codeman7 committed:
Address feedback

Address feedback and add test where applicable. This adds a testing interface for MDCRippleLayer and renames a fake class to fit it's use case better. Also this updates the path when the rippleRadius is set.

material-automation commented on components/Ripple/tests/unit/MDCRippleLayerTests.m:
```diff
@@ -32,6 +40,12 @@ - (void)addAnimation:(CAAnimation *)anim forKey:(NSString *)key {
 
+- (void)setPathFromRadii {
+  [super setPathFromRadii];
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/Ripple/tests/unit/MDCRippleLayerTests.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

codeman7 committed:
Clangy clang

codeman7 committed:
Remove setter

Since startRippleAtPoint calls setPathFromRadii we do not need the setter. Additionally if the setter is there and a rippleLayer is already active, in a selected state, this could cause unexpected behavior.

codeman7 committed:
Rename property

This renames the property based on feedback. Additionally this updates the function name to match the style guide

codeman7 committed:
Clang

material-automation commented on components/Ripple/src/private/MDCRippleLayer.m:
```diff
@@ -28,26 +28,29 @@
-  CGRect ovalRect =
-      CGRectMake(CGRectGetMidX(self.bounds) - _rippleRadius,
-                 CGRectGetMidY(self.bounds) - _rippleRadius, _rippleRadius * 2, _rippleRadius * 2);
+  CGFloat radius = self.maxRippleRadius > 0 ? self.maxRippleRadius : GetDefaultRippleRadius(self.bounds);
```
clang-format suggested the following change:


```
-  CGFloat radius = self.maxRippleRadius > 0 ? self.maxRippleRadius : GetDefaultRippleRadius(self.bounds);
+  CGFloat radius =
+      self.maxRippleRadius > 0 ? self.maxRippleRadius : GetDefaultRippleRadius(self.bounds);
```

<!-- identifier: clang-format -->
<!-- File: components/Ripple/src/private/MDCRippleLayer.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

codeman7 committed:
Remove testing interfaces

codeman7 committed:
Rename maxRippleRadius to maximumRadius
